platforms/allwinner-d1/d1-core/src/dmac/descriptor.rs

```diff
@@ -8,299 +8,389 @@
         }
-        if let Some(link) = value.link {
-            let link = link as usize;
-            if (link & 0b11) != 0 {
-                return Err(());
-            }
+    }
+
+    pub fn dest_data_width(self, val: DataWidth) -> Self {
+        Self {
+            cfg: self.cfg.with(Configuration::DEST_DATA_WIDTH, val),
+            ..self
         }
+    }
 
-        let mut descriptor = Descriptor {
-            configuration: 0,
-            source_address: 0,
-            destination_address: 0,
-            byte_counter: 0,
-            parameter: 0,
-            link: 0,
-        };
+    pub fn bmode_sel(self, val: BModeSel) -> Self {
+        Self {
+            cfg: self.cfg.with(Configuration::BMODE_SEL, val),
+            ..self
+        }
+    }
 
-        // Set source
-        descriptor.source_address = source as u32;
-        //                        332222222222 11 11 11111100 00000000
-        //                        109876543210 98 76 54321098 76543210
-        descriptor.parameter &= 0b111111111111_11_00_11111111_11111111;
-        descriptor.parameter |= (((source >> 32) & 0b11) << 16) as u32;
+    pub const fn link(self, link: Option<*const ()>) -> Self {
+        Self { link, ..self }
+    }
 
-        // Set dest
-        descriptor.destination_address = destination as u32;
-        //                        332222222222 11 11 11111100 00000000
-        //                        109876543210 98 76 54321098 76543210
-        descriptor.parameter &= 0b111111111111_00_11_11111111_11111111;
-        descriptor.parameter |= (((destination >> 32) & 0b11) << 18) as u32;
-
-        descriptor.byte_counter = value.byte_counter as u32;
-
-        // Set configuration
-        descriptor.configuration |= value.bmode.to_desc_bits();
-        descriptor.configuration |= value.dest_width.to_desc_bits_dest();
-        descriptor.configuration |= value.dest_addr_mode.to_desc_bits_dest();
-        descriptor.configuration |= value.dest_block_size.to_desc_bits_dest();
-        descriptor.configuration |= value.dest_drq_type.to_desc_bits();
-        descriptor.configuration |= value.src_data_width.to_desc_bits_src();
-        descriptor.configuration |= value.src_addr_mode.to_desc_bits_src();
-        descriptor.configuration |= value.src_block_size.to_desc_bits_src();
-        descriptor.configuration |= value.src_drq_type.to_desc_bits();
-
-        if let Some(link) = value.link {
-            descriptor.link = link as u32;
-            // We already verified above the low bits of `value.link` are clear,
-            // no need to re-mask the current state of `descriptor.link`.
-            descriptor.link |= ((link as usize >> 32) as u32) & 0b11
-        } else {
-            descriptor.end_link();
+    pub const fn wait_clock_cycles(self, wait_clock_cycles: u8) -> Self {
+        Self {
+            wait_clock_cycles,
+            ..self
         }
-
-        Ok(descriptor)
     }
-}
 
-// DescriptorConfig
+    pub fn build(
```

Comment:
would it make sense to have an outer method that is generic over `<S, D>` types so the caller doesn't have to do pointer shenanigans?

so rename THIS function as `build_inner`, then something like:

```rust
fn build<S, D>(self, src: &S, dst: &mut D, byte_counter: u32) -> Result<...> {
    let src = src.as_ptr().cast::<()>();
    let dst = dst.as_mut_ptr().cast::<()>();
    build_inner(src, dst, byte_counter)
}
```

just as a little QoL thing.

Reply:
hmm, that's not a bad idea, we could do something like that. potentially we could even have a version where the destination is taken as a `&mut [MaybeUninit<u8>]` and the length is used to generate `byte_counter`...

Reply:
i wonder if `dst` in this should actually be `Pin<&mut D>` so that we can prove we're not going to move the pointed buffer out from under the caller? maybe that's not worth it though idk.

Reply:
hmm, i tried adding a function like this, but i'm not sure if it's worth it, since it actually makes both existing uses of `build` _more_ complex, as we need to turn the raw pointers to the various MMIO dest registers into `&mut whatever`s, just so we can pass them back to the function that turns them into raw pointers:

```patch
diff --git a/platforms/allwinner-d1/d1-core/src/dmac/descriptor.rs b/platforms/allwinner-d1/d1-core/src/dmac/descriptor.rs
index 9090920..583456f 100644
--- a/platforms/allwinner-d1/d1-core/src/dmac/descriptor.rs
+++ b/platforms/allwinner-d1/d1-core/src/dmac/descriptor.rs
@@ -272,7 +272,18 @@ impl DescriptorBuilder {
         }
     }
 
-    pub fn build(
+    pub fn build<S, D>(
+        self,
+        source: &S,
+        destination: &mut D,
+        byte_counter: u32,
+    ) -> Result<Descriptor, InvalidDescriptor> {
+        let source = source as *const _ as *const ();
+        let destination = destination as *mut _ as *mut ();
+        self.build_raw(source, destination, byte_counter)
+    }
+
+    pub fn build_raw(
         self,
         source: *const (),
         destination: *mut (),
diff --git a/platforms/allwinner-d1/d1-core/src/drivers/spim.rs b/platforms/allwinner-d1/d1-core/src/drivers/spim.rs
index c853f43..9806a0a 100644
--- a/platforms/allwinner-d1/d1-core/src/drivers/spim.rs
+++ b/platforms/allwinner-d1/d1-core/src/drivers/spim.rs
@@ -118,9 +118,7 @@ impl SpiSenderServer {
             .spawn(async move {
                 let spi = unsafe { &*SPI_DBI::PTR };
 
-                let txd_ptr: *mut u32 = spi.spi_txd.as_ptr();
-                let txd_ptr: *mut u8 = txd_ptr.cast();
-                let txd_ptr: *mut () = txd_ptr.cast();
+                let txd = unsafe { &mut *(spi.spi_txd.as_ptr() as *mut _) };
 
                 let descr_cfg = Descriptor::builder()
                     .dest_data_width(DataWidth::Bit8)
@@ -137,20 +135,23 @@ impl SpiSenderServer {
                     let Message { msg, reply } = reqs.next_request().await;
                     let SpiSenderRequest::Send(ref payload) = msg.body;
 
-                    let len = payload.as_slice().len();
+                    let payload = payload.as_slice();
+                    // TODO(eliza): we should probably respond with an error if
+                    // someone tries to send a payload whose length is > u32::MAX...;
+                    let len = payload.len() as u32;
 
                     spi.spi_bcc.modify(|_r, w| {
                         // "Single Mode Transmit Counter" - the number of bytes to send
-                        w.stc().variant(len as u32);
+                        w.stc().variant(len);
                         w
                     });
                     spi.spi_mbc.modify(|_r, w| {
                         // Master Burst Counter
-                        w.mbc().variant(len as u32);
+                        w.mbc().variant(len);
                         w
                     });
                     spi.spi_mtc.modify(|_r, w| {
-                        w.mwtc().variant(len as u32);
+                        w.mwtc().variant(len);
                         w
                     });
                     // Start transfer
@@ -160,7 +161,7 @@ impl SpiSenderServer {
                     });
 
                     let descriptor = descr_cfg
-                        .build(payload.as_slice().as_ptr().cast(), txd_ptr, len as u32)
+                        .build(&payload[0], txd, len)
                         .expect("failed to build SPI1_TX DMA descriptor");
 
                     // start the DMA transfer.
diff --git a/platforms/allwinner-d1/d1-core/src/drivers/uart.rs b/platforms/allwinner-d1/d1-core/src/drivers/uart.rs
index 2b9ba2f..db9ad9e 100644
--- a/platforms/allwinner-d1/d1-core/src/drivers/uart.rs
+++ b/platforms/allwinner-d1/d1-core/src/drivers/uart.rs
@@ -124,12 +124,12 @@ impl D1Uart {
         loop {
             let rx = cons.read_grant().await;
             let len = rx.len();
-            let thr_addr = unsafe { &*UART0::PTR }.thr() as *const _ as *mut ();
+            let thr = unsafe { &mut *((*UART0::PTR).thr() as *const _ as *mut u32) };
 
             let rx_sli: &[u8] = &rx;
 
             let descriptor = descr_cfg
-                .build(rx_sli.as_ptr().cast(), thr_addr, rx_sli.len() as u32)
+                .build(&rx_sli[0], thr, rx_sli.len() as u32)
                 .expect("failed to build UART0 DMA transfer descriptor");
 
             // start the DMA transfer.

```
wdyt?

Reply:
...aaand `rustc` actually does not like that code at all:
```
error: casting `&T` to `&mut T` is undefined behavior, even if the reference is unused, consider instead using an `UnsafeCell`
   --> platforms/allwinner-d1/d1-core/src/drivers/uart.rs:127:32
    |
127 |             let thr = unsafe { &mut *((*UART0::PTR).thr() as *const _ as *mut u32) };
    |                                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: `#[deny(invalid_reference_casting)]` on by default
```
so we probably need to actually just use the raw-ptr `build_raw` regardless...

Reply:
as of 42671b95637b383315f8933ccc2012ed0158a82a, i've figured out a nice way to do this. :)